File: src/checks.ts
import { execSync } from 'node:child_process'
import type { Checks } from './config.ts'

type CheckResult = {
  lint: boolean
  test: boolean
}

export async function runLintAndTest(checks: Checks): Promise<CheckResult> {
  let lint = true
  let test = true

  if (checks.lintCmd) {
    lint = await runCmd(checks.lintCmd)
  }

  if (checks.testCmd) {
    test = await runCmd(checks.testCmd)
  }

  return {
    lint,
    test,
  }
}

export async function runCmd(cmd: string): Promise<boolean> {
  try {
    execSync(cmd, {
      encoding: 'utf-8',
    })

    return true
  } catch {
    return false
  }
}

File: matrix-go/checks.go
package main

import (
	"os"
	"os/exec"
)

func runCmd(cmd string) bool {
	c := exec.Command("sh", "-c", cmd)
	c.Stdout = os.Stdout
	c.Stderr = os.Stderr
	return c.Run() == nil
}

func runChecks(c *Checks) (lint, test bool) {
	lint = true
	test = true
	if c.LintCmd != "" {
		lint = runCmd(c.LintCmd)
	}
	if c.TestCmd != "" {
		test = runCmd(c.TestCmd)
	}
	return
}
